src/query/plan/rule_based_planner.cpp

```diff
@@ -516,14 +517,30 @@ bool HasBoundFilterSymbols(const std::unordered_set<Symbol> &bound_symbols, cons
 
 Expression *ExtractFilters(const std::unordered_set<Symbol> &bound_symbols, Filters &filters, AstStorage &storage) {
   Expression *filter_expr = nullptr;
+  std::vector<FilterInfo> and_joinable_filters{};
   for (auto filters_it = filters.begin(); filters_it != filters.end();) {
     if (HasBoundFilterSymbols(bound_symbols, *filters_it)) {
-      filter_expr = impl::BoolJoin<AndOperator>(storage, filter_expr, filters_it->expression);
+      and_joinable_filters.emplace_back(*filters_it);
       filters_it = filters.erase(filters_it);
     } else {
       filters_it++;
     }
   }
+  // Idea here is to join filters in a way
+  // that pattern filter ( exists() ) is at the end
+  // so if any of the AND filters before
+  // evaluate to false we don't need to
+  // evaluate pattern ( exists() ) filter
+  auto split =
+      std::partition(and_joinable_filters.begin(), and_joinable_filters.end(),
+                     [](const FilterInfo &filter_info) { return filter_info.type != FilterInfo::Type::Pattern; });
+  for (auto it = and_joinable_filters.begin(); it != split; ++it) {
+    filter_expr = impl::BoolJoin<AndOperator>(storage, filter_expr, it->expression);
+  }
+  for (; split != and_joinable_filters.end(); ++split) {
+    filter_expr = impl::BoolJoin<AndOperator>(storage, filter_expr, split->expression);
+  }
```

Comment:
Why looping here 2 times? Isn't `std::partition` operating on the `and_joinable_filters` and then you can just go over them and `impl::BoolJoin<AndOperator>`?

https://en.cppreference.com/w/cpp/algorithm/partition

Reply:
Yes I also think you can just do in one loop since there is no iterator which you want to avoid, you want to connect all of them

Reply:
My bad